api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -744,54 +748,60 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
 
   @Override
   public ResponseEntity<FileDetail> renameNotebook(String workspace, String workspaceName, NotebookRename rename) {
-    String bucket = fireCloudService.getWorkspace(workspace, workspaceName)
-      .getWorkspace()
-      .getBucketName();
-    String origPath = NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + rename.getName();
     String newName = rename.getNewName();
     if (!newName.matches("^.+\\.ipynb")) {
       newName = newName + ".ipynb";
     }
-    String newPath = NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + newName;
-    BlobId blobId = BlobId.of(bucket, origPath);
-    cloudStorageService.copyBlob(blobId, BlobId.of(bucket, newPath));
-    cloudStorageService.deleteBlob(blobId);
-    FileDetail fileDetail = new FileDetail();
-    fileDetail.setName(newName);
-    fileDetail.setPath("gs://" + bucket + "/" + NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + newName);
-    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
-    fileDetail.setLastModifiedTime(now.getTime());
+    FileDetail fileDetail = performNotebookOperation("rename", workspace, workspaceName, rename.getName(), newName);
     return ResponseEntity.ok(fileDetail);
   }
 
   @Override
   public ResponseEntity<FileDetail> cloneNotebook(String workspace, String workspaceName, String notebookName) {
-    String bucket = fireCloudService.getWorkspace(workspace, workspaceName)
-      .getWorkspace()
-      .getBucketName();
-    String origPath = NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + notebookName;
     String newName = notebookName.replaceAll("\\.ipynb", " ") + "Clone.ipynb";
-    String newPath = NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + newName;
-    BlobId blobId = BlobId.of(bucket, origPath);
-    cloudStorageService.copyBlob(blobId, BlobId.of(bucket, newPath));
-    FileDetail fileDetail = new FileDetail();
-    fileDetail.setName(newName);
-    fileDetail.setPath("gs://" + bucket + "/" + newPath);
-    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
-    fileDetail.setLastModifiedTime(now.getTime());
+    FileDetail fileDetail = performNotebookOperation("copy", workspace, workspaceName, notebookName, newName);
     return ResponseEntity.ok(fileDetail);
   }
 
   @Override
   public ResponseEntity<EmptyResponse> deleteNotebook(String workspace, String workspaceName, String notebookName) {
+    performNotebookOperation("delete", workspace, workspaceName, notebookName, "");
+    return ResponseEntity.ok(new EmptyResponse());
+  }
+
+  private FileDetail performNotebookOperation(String operation, String workspace, String workspaceName, String notebookName, String newName) {
```

Comment:
I like this a lot. Good refactor. (Always happy to see extra code go away)

An alternative way to do this could also be a 'notebookCloneOperation' and a 'notebookDeleteOperation'? Then rename would call Clone and then Delete, clone would just call Clone, and delete just delete?

Reply:
Ooooh hm, I actually like that even better. Yup! Ok going to do that!

Reply:
Hmmm, or maybe variation of that... a lot of the setup for both operations is the same....